src/java.base/share/classes/java/lang/classfile/TypeKind.java

```diff
@@ -239,30 +241,38 @@ public static TypeKind fromDescriptor(CharSequence s) {
     }
 
     /**
-     * {@return the type associated with the specified field descriptor}
-     * @param descriptor the field descriptor
+     * {@return the type associated with the specified ClassDesc}
+     * @param desc the ClassDesc
      */
-    public static TypeKind from(TypeDescriptor.OfField<?> descriptor) {
-        return descriptor.isPrimitive() // implicit null check
-                ? fromDescriptor(descriptor.descriptorString())
-                : REFERENCE;
+    public static TypeKind from(ClassDesc desc) {
+        if (desc == null      ) throw new NullPointerException();
+        if (desc == CD_boolean) return BOOLEAN;
+        if (desc == CD_byte   ) return BYTE;
+        if (desc == CD_char   ) return CHAR;
+        if (desc == CD_int    ) return INT;
+        if (desc == CD_long   ) return LONG;
+        if (desc == CD_short  ) return SHORT;
+        if (desc == CD_float  ) return FLOAT;
+        if (desc == CD_double ) return DOUBLE;
+        if (desc == CD_void   ) return VOID;
+        else                    return REFERENCE;
```

Comment:
This can short‑circuit when `desc` is not a `PrimitiveClassDescImpl`:
```suggestion
        if (desc instanceof PrimitiveClassDescImpl) {
            if (desc == CD_boolean) return BOOLEAN;
            if (desc == CD_byte   ) return BYTE;
            if (desc == CD_char   ) return CHAR;
            if (desc == CD_int    ) return INT;
            if (desc == CD_long   ) return LONG;
            if (desc == CD_short  ) return SHORT;
            if (desc == CD_float  ) return FLOAT;
            if (desc == CD_double ) return DOUBLE;
            if (desc == CD_void   ) return VOID;
        }
        return REFERENCE;
```